File: array-hash/27-Remove-Element.py
class Solution:
    def removeElement(self, nums: List[int], val: int) -> int:
        left, right = 0, len(nums) - 1
    
        while left < right:
            if nums[left] == val:
                if nums[left] != nums[right]:
                    nums[left], nums[right] = nums[right], nums[left]
                    left += 1
                right -= 1
            else:
                left += 1

        k = 0
        for i in range(len(nums)):
            if nums[i] != val:
                k += 1
            else:
                break
        
        return k

File: reference/array-hash/27-Remove-Element.go
func removeElement(nums []int, val int) int {
	left, right := 0, len(nums)-1
	for left < right {
		if nums[left] == val {
			if nums[left] != nums[right] {
				nums[left], nums[right] = nums[right], nums[left]
				left++
			}
			right--
		} else {
			left++
		}
	}

	k := 0
	for i := 0; i < len(nums); i++ {
		if nums[i] != val {
			k++
		} else {
			break
		}
	}
	return k
}

// another solution
func removeElement(nums []int, val int) int {
	k := 0
	for i := 0; i < len(nums); i++ {
		if nums[i] != val {
			nums[k] = nums[i]
			k++
		}
	}
	return k
}